From: Neil Horman <nhorman@openssl.org>
Date: Tue, 25 Aug 2026 14:23:50 -0400
Subject: don't double count full databgram length on unvalidated connections

If a connection is coalescing frames, we pass through
ossl_quic_handle_frames multiple times, each time accounting the full
datagram length to the connection, erroneously amplifying our
unvalidated credit.

Fix it by moving where we account unvalidated credit.  If we move the
adding of unvalidated credit to port_default_packet_handler, we can add
the datagram length to the channels unvalidated credit before it gets
broken up into multiple OSSL_QRX_PKT structures.

Fixes CVE-2026-35191
---
 ssl/quic/quic_port.c      | 2 ++
 ssl/quic/quic_rx_depack.c | 2 --
 2 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/ssl/quic/quic_port.c b/ssl/quic/quic_port.c
index aad9c3a5b3d5..6ddc1322897a 100644
--- a/ssl/quic/quic_port.c
+++ b/ssl/quic/quic_port.c
@@ -1510,6 +1510,7 @@ static void port_default_packet_handler(QUIC_URXE *e, void *arg,
         && ossl_quic_lcidm_lookup(port->lcidm, dcid, NULL,
             (void **)&ch)) {
         assert(ch != NULL);
+        ossl_quic_tx_packetiser_add_unvalidated_credit(ch->txp, e->data_len);
         ossl_quic_channel_inject(ch, e);
         return;
     }
@@ -1721,6 +1722,7 @@ static void port_default_packet_handler(QUIC_URXE *e, void *arg,
          * Time to reinject packets from qrx to channel before
          * qrx will be destroyed here.
          */
+        ossl_quic_tx_packetiser_add_unvalidated_credit(new_ch->txp, e->data_len);
         while (ossl_qrx_read_pkt(qrx_src, &qrx_pkt) == 1)
             ossl_quic_channel_inject_pkt(new_ch, qrx_pkt);
         ossl_qrx_update_pn_space(qrx_src, new_ch->qrx);
diff --git a/ssl/quic/quic_rx_depack.c b/ssl/quic/quic_rx_depack.c
index 7ab59f01a1cd..0ead9097016e 100644
--- a/ssl/quic/quic_rx_depack.c
+++ b/ssl/quic/quic_rx_depack.c
@@ -1486,8 +1486,6 @@ int ossl_quic_handle_frames(QUIC_CHANNEL *ch, OSSL_QRX_PKT *qpacket)
      */
     if (enc_level == QUIC_ENC_LEVEL_HANDSHAKE)
         ossl_quic_tx_packetiser_set_validated(ch->txp);
-    else
-        ossl_quic_tx_packetiser_add_unvalidated_credit(ch->txp, dgram_len);
 
     /* Now that special cases are out of the way, parse frames */
     if (!PACKET_buf_init(&pkt, qpacket->hdr->data, qpacket->hdr->len)
